File: app/src/main/java/net/veldor/flibustaloader/parsers/GenresParser.java
package net.veldor.flibustaloader.parsers;

import android.util.Log;

import net.veldor.flibustaloader.App;
import net.veldor.flibustaloader.selections.Genre;

import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

import java.util.ArrayList;

import javax.xml.xpath.XPath;
import javax.xml.xpath.XPathConstants;
import javax.xml.xpath.XPathExpressionException;

class GenresParser {
    public static ArrayList parse(NodeList entries, XPath xPath) {
        ArrayList<Genre> result = new ArrayList<Genre>();
        Node entry;
        Genre genre;
        int counter = 0;
        int entriesLength = entries.getLength();
        int handledEntryCounter = 0;
        while ((entry = entries.item(counter)) != null) {
            ++handledEntryCounter;
            App.getInstance().mLoadAllStatus.postValue("Обрабатываю жанр " + handledEntryCounter + " из " + entriesLength);
            genre = new Genre();
            try {
                genre.label = ((Node) xPath.evaluate("./title", entry, XPathConstants.NODE)).getTextContent();
                genre.term = ((Node) xPath.evaluate("./link", entry, XPathConstants.NODE)).getAttributes().getNamedItem("href").getTextContent();
            } catch (XPathExpressionException e) {
                e.printStackTrace();
            }
            result.add(genre);
            counter++;
        }
        // отправлю обработанные результаты
        Log.d("surprise", "GenresParser parse: founded genres: " + result.size());
        return result;
    }
}

File: app/src/main/java/net/veldor/flibustaloader/parsers/GenresParser.kt
package net.veldor.flibustaloader.parsers

import net.veldor.flibustaloader.App
import net.veldor.flibustaloader.selections.Genre
import android.util.Log
import org.w3c.dom.Node
import org.w3c.dom.NodeList
import java.util.ArrayList
import javax.xml.xpath.XPath
import javax.xml.xpath.XPathConstants
import javax.xml.xpath.XPathExpressionException

internal object GenresParser {
    fun parse(entries: NodeList, xPath: XPath): ArrayList<*> {
        val result = ArrayList<Genre>()
        var entry: Node?
        var genre: Genre
        var counter = 0
        val entriesLength = entries.length
        var handledEntryCounter = 0
        while (entries.item(counter).also { entry = it } != null) {
            ++handledEntryCounter
            genre = Genre()
            try {
                genre.label =
                    (xPath.evaluate("./title", entry, XPathConstants.NODE) as Node).textContent
                genre.term = (xPath.evaluate(
                    "./link",
                    entry,
                    XPathConstants.NODE
                ) as Node).attributes.getNamedItem("href").textContent
            } catch (e: XPathExpressionException) {
                e.printStackTrace()
            }
            result.add(genre)
            counter++
        }
        // отправлю обработанные результаты
        Log.d("surprise", "GenresParser parse: founded genres: " + result.size)
        return result
    }
}